From ff88b8f06fe81e1dadba1048ac6fb626b9403f74 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Thu, 4 May 2023 16:08:11 +0200
Subject: [PATCH 7/9] Add support for userdata apk packages

Allow user to configure apk cache and apk repositories and install
packages.
---
 lib/tiny-cloud/init  | 52 ++++++++++++++++++++++++++++++
 tests/init-main.test | 75 +++++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 126 insertions(+), 1 deletion(-)

diff --git a/lib/tiny-cloud/init b/lib/tiny-cloud/init
index 0ac88f8..c655e27 100644
--- a/lib/tiny-cloud/init
+++ b/lib/tiny-cloud/init
@@ -20,6 +20,9 @@ INIT_ACTIONS_MAIN="
     save_userdata
     set_hostname
     userdata_ntp
+    userdata_apk_cache
+    userdata_apk_repositories
+    userdata_packages
     set_ssh_keys
 "
 INIT_ACTIONS_FINAL="
@@ -180,6 +183,55 @@ init__userdata_ntp() {
     fi
 }
 
+init__userdata_apk_cache() {
+    local cache="$(get_userdata apk cache)"
+    if [ -z "$cache" ]; then
+        return
+    fi
+    mkdir -p "$ROOT/$cache"
+    # make link relative
+    case "$cache" in
+        /*) cache="../..$cache";;
+    esac
+    mkdir -p "$ROOT"/etc/apk
+    ln -sf "$cache" "$ROOT"/etc/apk/cache
+}
+
+init__userdata_apk_repositories() {
+    local repositories="$(get_userdata apk repositories)"
+    mkdir -p "$ROOT"/etc/apk
+    for r in $repositories; do
+        local baseurl="$(get_userdata apk repositories $r base_url)"
+        local repos="$(get_userdata apk repositories $r repos)"
+        local version="$(get_userdata apk repositories $r version)"
+        if [ -z "$version" ]; then
+            local version_id=$( . "$ROOT"/etc/os-release 2>/dev/null && echo "$VERSION_ID")
+            case "$version_id" in
+                edge*|*_alpha*) version="edge";;
+                [0-9]*.[0-9]*.[0-9]*) version="v${version_id%.*}";;
+            esac
+        fi
+        if [ -n "$version" ] && [ "$version" != "." ] && [ "$version" != "/" ]; then
+            baseurl="${baseurl%/}/$version"
+        fi
+        for repo in $repos; do
+            local uri="${baseurl%/}/$(get_userdata apk repositories $r repos $repo)"
+            add_once "$ROOT"/etc/apk/repositories "$uri"
+        done
+    done
+}
+
+init__userdata_packages() {
+    local packages="$(get_userdata packages)"
+    local pkgs=
+    for i in $packages; do
+        pkgs="$pkgs $(get_userdata packages $i)"
+    done
+    if [ -n "$pkgs" ]; then
+        $MOCK apk add $pkgs
+    fi
+}
+
 ### init-final functions
 
 init__run_userdata() {
diff --git a/tests/init-main.test b/tests/init-main.test
index 05cda06..68bf897 100755
--- a/tests/init-main.test
+++ b/tests/init-main.test
@@ -13,7 +13,12 @@ init_tests \
 	save_userdata_compressed \
 	userdata_ntp \
 	userdata_ntp_busybox \
-	userdata_ntp_openntpd
+	userdata_ntp_openntpd \
+	userdata_apk_cache \
+	userdata_apk_repositories \
+	userdata_apk_repositories_version \
+	userdata_apk_repositories_version_auto_edge \
+	userdata_packages
 
 
 set_nocloud_meta() {
@@ -128,3 +133,71 @@ userdata_ntp_openntpd_body() {
 		-o match:"rc-service .*openntpd" \
 		sh -c ". \"$lib\"; init__userdata_ntp"
 }
+
+userdata_apk_cache_body() {
+	fake_userdata_nocloud <<-EOF
+		apk:
+		  cache: /var/cache/apk
+	EOF
+	CLOUD=nocloud atf_check \
+		sh -c ". \"$lib\"; init__userdata_apk_cache"
+	atf_check -o match:"$PWD/var/cache/apk" readlink -f etc/apk/cache
+}
+
+userdata_apk_repositories_body() {
+	fake_userdata_nocloud <<-EOF
+		apk:
+		  repositories:
+		    - base_url: /srv/packages
+		      repos: [ "main", "community" ]
+	EOF
+	CLOUD=nocloud atf_check \
+		sh -c ". \"$lib\"; init__userdata_apk_repositories"
+	atf_check -o match:"^/srv/packages/main$" \
+		-o match:"^/srv/packages/community$" \
+		cat etc/apk/repositories
+}
+
+userdata_apk_repositories_version_body() {
+	fake_userdata_nocloud <<-EOF
+		apk:
+		  repositories:
+		    - base_url: https://cdn.alpinelinux.org/
+		      version: edge
+		      repos: [ "main", "community" ]
+	EOF
+	CLOUD=nocloud atf_check \
+		sh -c ". \"$lib\"; init__userdata_apk_repositories"
+	atf_check -o match:"^https://cdn.alpinelinux.org/edge/main$" \
+		-o match:"^https://cdn.alpinelinux.org/edge/community$" \
+		cat etc/apk/repositories
+}
+
+userdata_apk_repositories_version_auto_edge_body() {
+	fake_userdata_nocloud <<-EOF
+		apk:
+		  repositories:
+		    - base_url: https://cdn.alpinelinux.org/
+		      repos: [ "main", "community" ]
+	EOF
+	mkdir -p etc
+	echo "VERSION_ID=3.18_alpha20230329" > etc/os-release
+
+	CLOUD=nocloud atf_check \
+		sh -c ". \"$lib\"; init__userdata_apk_repositories"
+	atf_check -o match:"^https://cdn.alpinelinux.org/edge/main$" \
+		-o match:"^https://cdn.alpinelinux.org/edge/community$" \
+		cat etc/apk/repositories
+}
+
+userdata_packages_body() {
+	fake_userdata_nocloud <<-EOF
+		packages:
+		  - tmux
+		  - vim
+	EOF
+	CLOUD=nocloud atf_check \
+		-o match:"apk add .*tmux" \
+		-o match:"apk add .*vim" \
+		sh -c ". \"$lib\"; init__userdata_packages"
+}
-- 
2.40.1

